drivers/can/stm32_can.c

```diff
@@ -841,51 +841,66 @@ static const struct can_driver_api can_api_funcs = {
 	.detach = can_stm32_detach
 };
 
-#ifdef CONFIG_CAN_1
-
-static void config_can_1_irq(CAN_TypeDef *can);
-
-static const struct can_stm32_config can_stm32_cfg_1 = {
-	.can = (CAN_TypeDef *)CONFIG_CAN_1_BASE_ADDRESS,
-	.bus_speed = CONFIG_CAN_1_BUS_SPEED,
-	.swj = CONFIG_CAN_1_SJW,
-	.prop_bs1 = CONFIG_CAN_1_PROP_SEG_PHASE_SEG1,
-	.bs2 = CONFIG_CAN_1_PHASE_SEG2,
-	.pclken = {
-		.enr = CONFIG_CAN_1_CLOCK_BITS,
-		.bus = CONFIG_CAN_1_CLOCK_BUS,
-	},
-	.config_irq = config_can_1_irq
-};
-
-static struct can_stm32_data can_stm32_dev_data_1;
-
-DEVICE_AND_API_INIT(can_stm32_1, CONFIG_CAN_1_NAME, &can_stm32_init,
-		    &can_stm32_dev_data_1, &can_stm32_cfg_1,
-		    POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
-		    &can_api_funcs);
-
-static void config_can_1_irq(CAN_TypeDef *can)
-{
-	LOG_DBG("Enable CAN1 IRQ");
-#ifdef CONFIG_SOC_SERIES_STM32F0X
-	IRQ_CONNECT(CONFIG_CAN_1_IRQ, CONFIG_CAN_1_IRQ_PRIORITY, can_stm32_isr,
-		    DEVICE_GET(can_stm32_1), 0);
-	irq_enable(CONFIG_CAN_1_IRQ);
-#else
-	IRQ_CONNECT(CONFIG_CAN_1_IRQ_RX0, CONFIG_CAN_1_IRQ_PRIORITY,
-		    can_stm32_rx_isr, DEVICE_GET(can_stm32_1), 0);
-	irq_enable(CONFIG_CAN_1_IRQ_RX0);
-
-	IRQ_CONNECT(CONFIG_CAN_1_IRQ_TX, CONFIG_CAN_1_IRQ_PRIORITY,
-		    can_stm32_tx_isr, DEVICE_GET(can_stm32_1), 0);
-	irq_enable(CONFIG_CAN_1_IRQ_TX);
-
-	IRQ_CONNECT(CONFIG_CAN_1_IRQ_SCE, CONFIG_CAN_1_IRQ_PRIORITY,
-		    can_stm32_tx_isr, DEVICE_GET(can_stm32_1), 0);
-	irq_enable(CONFIG_CAN_1_IRQ_SCE);
-#endif
-	can->IER |= CAN_IT_TME | CAN_IT_ERR | CAN_IT_FMP0 | CAN_IT_FMP1;
-}
-
-#endif /*CONFIG_CAN_1*/
+/**
+ * @code{.codegen}
+ * codegen.import_module('devicedeclare')
+ *
+ * device_configs = ['CONFIG_CAN_1', ]
+ * driver_names = ['CAN_1', ]
```

Comment:
As mentioned several times, this should not be present in API.
This information should be extracted from device tree and drivers <> EDTS binding should be done by 'compatible' property. This is literally its main and only use.
cf #8561

Reply:
As mentionend several times to identify a specific driver instance the compatible is not enough. There are usual several driver instantiations for one compatible. So you need something more to identify a single one and use the 'label' property for that. But if you use the 'label' property in addition to the 'compatible' you find out that the 'label' alone is unique und you really do not need the 'compatible'. Taking the 'compatible' and the 'label' is just a waste of computing.